prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+{
+	int optval;
+	optval = 1;
+
+	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt reuseaddr failed\n");
+
+	if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt nodelay failed\n");
+
+}
+
+static void *tcpx_ep_connect_handler(void *data)
+{
+	struct tcpx_conn_handle *handle;
+	struct fi_eq_cm_entry *eq_entry;
+	struct tcpx_ep *tcpx_ep = (struct tcpx_ep *) data;
+
+	handle = tcpx_ep->handle;
+
+	eq_entry  = calloc(1, sizeof(*eq_entry));
+	if (!eq_entry) {
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "cannot allocate memory\n");
+	  goto err;
+	}
+
+	eq_entry->fid = &tcpx_ep->util_ep.ep_fid.fid;
+	eq_entry->info = &tcpx_ep->info;
+
+	tcpx_set_sockopts(handle->conn_fd);
+	if (connect(handle->conn_fd,
+		    (struct sockaddr *)handle->serv_addr,
+		    sizeof(*handle->serv_addr))) {
+
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "connect failed : %s\n", strerror(errno));
+	  goto err1;
+	}
+	fi_fd_nonblock(handle->conn_fd);
+
+	/* report success to eq with FI_CONNECTED */
+	fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_CONNECTED,
+		    eq_entry, sizeof(*eq_entry), 0);
+
+
+
+	return NULL;
+err1:
+	free(eq_entry);
+err:
+	ofi_close_socket(handle->conn_fd);
+
+	/* report the error to eq */
+	fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
+		    eq_entry, sizeof(*eq_entry), 0);
+
+
+	return NULL;
+}
+
+
+static int tcpx_ep_connect(struct fid_ep *ep, const void *addr,
+		    const void *param, size_t paramlen)
+{
+	struct tcpx_ep *tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+	struct sockaddr_in *serv_addr = (struct sockaddr_in *) addr;
+	struct tcpx_conn_handle *handle = tcpx_ep->handle;
+	int ret = 0;
+
+	if (!addr)
+		return -FI_EINVAL;
+
+	handle->conn_fd = socket(AF_INET, SOCK_STREAM, 0);
+	if (handle->conn_fd < 0) {
+		ret = handle->conn_fd;
+		goto out;
+	}
+
+	tcpx_ep->handle->serv_addr = serv_addr;
+
+	if (pthread_create(&tcpx_ep->cm_thread, NULL,
+			   tcpx_ep_connect_handler, (void *) tcpx_ep)) {
+
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create connect thread\n");
+		goto err;
+	}
+
+
+	return FI_SUCCESS;
+err:
+	ofi_close_socket(handle->conn_fd);
+out:
+	free(handle);
+	return ret;
+
+}
+
+static int tcpx_ep_accept(struct fid_ep *ep, const void *param, size_t paramlen)
+{
+	struct fi_eq_cm_entry *eq_entry;
+	struct tcpx_ep *tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+
+	if (tcpx_ep->handle != NULL && tcpx_ep->handle->conn_fd != 9) {
+
+		eq_entry  = calloc(1, sizeof(*eq_entry));
+		if (!eq_entry) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"cannot allocate memory\n");
+			return -FI_ENOMEM;
+		}
+
+		eq_entry->fid = &ep->fid;
+		eq_entry->info = &tcpx_ep->info;
+
+		/* report FI_CONNECTED to eq  */
+		fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_CONNECTED, eq_entry,
+			    sizeof(*eq_entry), 0);
+		return 0;
+	}
+	return -FI_EINVAL;
```

Comment:
Reverse the above if statement to check for a failure and abort immediately, rather than having the entire successful code path indented inside of a if statement.